SSA/Core/MLIRSyntax/Parser.lean

```diff
@@ -0,0 +1,138 @@
+ -/
+def elabRegion (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Region 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Region 0) env region0Expr stx
+  return reg
+
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Op`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabOp (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Op 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Op 0) env op0Expr stx
+  return reg
+
+def mkParseFun (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (s : String) (env : Environment) : EIO ParseError ParseOutput := do
+  let syn : Except String Syntax := Parser.runParserCategory env syntaxcat s
+  match syn with
+    | .error msg => throw msg
+    | .ok syn => ntparser syn
+
+private def mkNonTerminalParser (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (env : Environment) (s : String) : EIO ParseError ParseOutput :=
+  let parseFun := mkParseFun syntaxcat ntparser
+  parseFun s env
+
+-- TODO: do we need this copy of the environment, or can I remove it from one of the two?
+def regionParser : @ParseFun (MLIR.AST.Region 0) :=
+  fun env : Lean.Environment => mkNonTerminalParser `mlir_region (elabRegion env) env
+
+private def parseFile (env: Lean.Environment)
+    (parser: @ParseFun ParseOutput)
+    (filepath: System.FilePath) : IO (Option ParseOutput) := do
+  let lines ← IO.FS.lines filepath
+  let fileStr := lines.foldl (λ s₁ s₂ => s₁ ++ "\n" ++ s₂) "" -- TODO: use intercalate.
+  let parsed ← EIO.toIO' <| parser env fileStr
+  match parsed with
+  | .ok parseOutput => return parseOutput
+  | .error msg => IO.println s!"Error parsing {filepath}:\n{msg}"; return none
+
+private def isFile (p: System.FilePath) : IO Bool := do
+  return (← p.metadata).type == IO.FS.FileType.file
+
+def runParser (parser : @ParseFun ParseOutput) (fileName : String) : IO (Option ParseOutput) := do
+  let executablePath : System.FilePath ← IO.appPath
+  let dotLakePath : System.FilePath :=
+    match executablePath.parent with
+      | .none => ".lake"
+      | .some binPath => match binPath.parent with
+        |.none => ".lake"
+        |.some buildPath => match buildPath.parent with
+          | .none => ".lake"
+          | .some dotLakePath => dotLakePath
+  let libPath : System.FilePath := dotLakePath / "build" / "lib"
+  -- TODO: can we get these from the lakefile?
+  let deps := ["mathlib", "std", "aesop", "Qq", "importGraph", "partax", "proofwidgets", "Cli"]
+  let depsPackages : List System.FilePath := deps.map fun depName =>
+    dotLakePath / "packages" / depName / ".lake" / "build" / "lib"
```

Comment:
This too feels fragile. Any chance we can as Mac if it's possible to use `Lake` as a library to achieve this?